### PR TITLE
Log aggregation

### PR DESCRIPTION
**Pull request by mhchia**
### What was wrong?
https://github.com/ethresearch/sharding-p2p-poc/issues/149

#### Features
Add data structures
- `map_log_enum_pattern`: a dictionary mapping Enum of the logs to their corresponding log patterns.
- `Event(time, log_type, logger_name, event_type, params)`: a log whose type is specified in `map_log_enum_pattern`.

Add functions
- `Network.get_events()`: returns a dictionary that maps each node to a tuple of `Event`, which happened in itself.

Fixes
- Fix `wait_for_log` to remove the ANSI color control codes to make parsing easier.

Add more tests
- Bunches of tests of `test_logs.py` and `test_log_aggregation.py`.
- Ensure all logs are emitted after all operations in the tests.

#### TODO
- [x] Parameters handler for each type of log
- [x] Fix the slow `Node.wait_fot_log()`
- [x] Unit-test for `discovershard`
- [x] `parse_logs` for each node
- [x] Unit-test for `get_events`

#### Issues
- [x] `wait_for_log` calls `grep` in the shell, whose regex is slightly different from the one in the Python module `re`. Need to modify something like `REGEX_LIST`, to ensure the regex also works in `grep`.
  - Note: there is a comparison table between linux grep and python, however only in Chinese: [link](http://blog.51cto.com/me2xp/1435095)

#### Cute Animal Picture

put a cute animal picture link inside the parentheses

### CODE REVIEW

File: test/simulation/simulation/logs.py [hunk 1]
```diff
@@ -1,13 +1,83 @@
...
+    LOG_BOOTSTRAP_FINISHED = auto()
+
+
+class OperationLogs(Enum):
+    LOG_RECEIVE_MSG = auto()
+
+
+# The regex for the list of elements: empty list, or 1 or more words, delimited by whitespaces
+# E.g. [], [1], [1 2], [shardCollations_2]
+REGEX_LIST = r'\[(?:|\w+(?: \w+)*)\]'
```
Comment (mhchia) [inline test/simulation/simulation/logs.py]: `\w` in `[]` and `(?:)` might not work in `grep`. Probably need to modify them

Comment (mhchia) [reply test/simulation/simulation/logs.py]: Sorry I'm wrong. [Non-capturing groups](https://www.regular-expressions.info/brackets.html) `(?:)` are supported in `egrep`.

**Comment by mhchia** [comment]
Ready for review. cc @NIC619

File: test/simulation/simulation/log_aggregation.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+)
+
+from .logs import (
+    EventHasNoParameter,
+    map_log_enum_pattern,
+    parse_event_params,
+)
+
+
+class NoMatchingPattern(Exception):
```
Comment (NIC619) [inline test/simulation/simulation/log_aggregation.py]: Should we move these to `exceptions.py` and import from there?

File: test/simulation/tests/test_node.py [hunk 3]
```diff
@@ -163,9 +171,17 @@ def test_add_peer(nodes):
     assert len(nodes[0].list_peer()) == 0
     assert len(nodes[1].list_peer()) == 0
     nodes[0].add_peer(nodes[1])
+    nodes[0].wait_for_log(
+        map_log_enum_pattern[RPCLogs.LOG_ADD_PEER_FMT],
+        0,
+    )
+    nodes[0].wait_for_log(
+        map_log_enum_pattern[RPCLogs.LOG_ADD_PEER_FINISHED],
```
Comment (NIC619) [inline test/simulation/tests/test_node.py:179]: If we wait for `*_FINISHED` logs, what's the reason to wait for `*_FMT` prior to `*_FINISHED`?

**Review by NIC619 — APPROVED** [review]
LGTM

File: test/simulation/simulation/log_aggregation.py (hunk 2 above)
Comment (mhchia) [reply test/simulation/simulation/log_aggregation.py]: Sure!

File: test/simulation/tests/test_node.py (hunk 3 above)
Comment (mhchia) [reply test/simulation/tests/test_node.py:179]: The difference between `*_FMT` and `*_FINISHED` is, there are arguments in `*_FMT`, while `*_FINISHED` not. I think `*_FMT` will be useful in this case, when we want to make sure an event with specific arguments has finished.

**Pull request merged**